webapp/lib/router.dart

```diff
@@ -1,28 +1,46 @@
 import 'dart:html';
+import 'controller.dart' as controller;
+
+class Route {
+  String path;
+  Function handler;
+
+  Route(this.path, this.handler);
+}
 
 class Router {
-  Map<String, Function> _routes;
+  Map<String, Route> _routes;
+  Route _authRoute;
 
   Router() {
     _routes = {};
   }
 
-  void addHandler(String route, Function callback) {
-    _routes[route] = callback;
+  void addHandler(Route route, {bool isAuthRoute = false}) {
+    _routes[route.path] = route;
+    if (isAuthRoute) {
+      _authRoute = route;
+    }
   }
 
-  void routeTo(String route) {
-    _loadView(route);
+  void listen() {
+    window.onPopState.listen((PopStateEvent event) => _loadView(window.location.hash));
   }
 
-  void _loadView(String route) {
-    if(_routes.containsKey(route)) {
-      _routes[route]();
-    }
+  void routeTo(String path) {
+    _loadView(path);
   }
 
-  void listen() {
-    window.onPopState.listen((_) => _loadView(window.location.hash));
+  void _loadView(String path) {
+    var targetRoute = _routes[path];
+    if (targetRoute != null) {
+      if (controller.signedInUser != null && targetRoute.path != _authRoute.path) {
+        targetRoute.handler();
+        window.location.hash = targetRoute.path;
+      } else if (controller.signedInUser == null) {
+        _authRoute.handler();
+        window.location.hash = targetRoute.path;
+      }
+    }
```

Comment:
this multi-level if/if/else if block is omitting some pages, and is a bit hard to follow. I suggest something like the following (not tested + needs other changes in the code + see comments inline):

```suggestion
    if (targetRoute == null) {
      _defaultRoute.handler(); // this needs adding in the same was as we add _authRoute, should point to the dashboard page
      window.location.hash = _defaultRoute.path;
      return;
    }
    if (controller.signedInUser == null) {
      _authRoute.handler();
      window.location.hash = targetRoute.path; // shouldn't this be _authRoute.path ?
      return;
    }
    targetRoute.handler();
    window.location.hash = targetRoute.path;
```

Reply:
Modified this a bit.